src/api/ElmParser.ts

```diff
@@ -0,0 +1,120 @@
+import { JSONPath } from "jsonpath-plus";
+
+class TextObj {
+  text: string;
+}
+class ChildObj {
+  children: any[] = [];
+
+  node_type: string;
+  ref_id: string;
+}
+class StatementObj {
+  children: ChildObj[] = [];
+  define_name: string;
+}
+class AnnotationObj {
+  statements: StatementObj[] = [];
+  identifier: any = {};
+}
+export const parse = (elmJson) => {
+  //This is what MITRE Cql Parser (ruby) returns
+  //  Statements is an array of nodes
+  // identifier is an object that contains "Id,Version" pairs for
+
+  const ret: AnnotationObj = new AnnotationObj();
+  const localid_to_type_map = {};
+  let nodes = [];
+
+  ["expression", "operand", "suchThat"].forEach((element) => {
+    //use a css selector to find all "field"s with either localId, xsi, or type
+    //nodes = elmJson.css(field + '[localId][xsi|type]')
+    //  * search for all nodes that contain "element, localID and type"
+    const path = `$.[?(@.${element})].[?(@.localId && @.type)]`;
+    const newNodes = JSONPath({ path: path, json: elmJson });
+    nodes.push(...newNodes);
+    nodes.forEach((node) => {
+      localid_to_type_map[node["localId"]] = node["type"];
+    });
+  });
+
+  //extract library identifier data
+  ret.identifier.id = JSONPath({
+    path: "$.library.identifier.id",
+    json: elmJson,
+  })[0];
+  ret.identifier.version = JSONPath({
+    path: "$.library.identifier.version",
+    json: elmJson,
+  })[0];
+  //all the define statements including functions
+  const annotations = JSONPath({
+    path: "$.library.statements.def.[?(@.annotation)]",
+    json: elmJson,
+  });
+
+  let index: number = 0;
+  annotations.forEach((annotation) => {
+    Object.keys(annotation).forEach((key) => {
```

Comment:
we probably do not need to iterate over all the props of definition. we can access the annotation property directly.

```
const annotation = definition.annotation;
    if (annotation) {
      ret.statements[0] = new StatementObj();
      ret.statements[0].children[index] = new ChildObj();
      parse_node(
        annotation,
        ret.statements[0].children[index++],
        localid_to_type_map
      );
    }
```